Next.JS - First build fails with "Module not found: Can't resolve './component.wyw-in-js.module.css'"

## Environment

- wyw-in-js version: 2.0.2
- custom processor: none
- Bundler (+ version): Turbopack Next.JS 16.2.9
- Node.js version: 24.15.0
- OS: MacOS

## Description

<!--
Describe your issue in detail.
Include screenshots if needed.
If this is a regression, let us know.
-->

After upgrading to @wyw-in-js/next to 2.0.2 Next.JS builds fail during the first build. 

I see that `.wyw-in-js.module.css` files are being generated during the first build, but Next.JS imports fails afterwards with messages such as:

```js
> Build error occurred
Error: Turbopack build failed with 2 errors:
./repo/src/component.style.ts:1:1
Module not found: Can't resolve './component.style.wyw-in-js.module.css'
> 1 | import "./component.style.wyw-in-js.module.css";
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  2 | export const wrapper = "w1ltldr3";
  3 | export const header = "h17yc8tj";
  4 | export const title = "t1snm9i1";
```

## Reproducible Demo

<!--
  Let us know how to reproduce the issue.
  Include a code sample or share a project that reproduces the issue.
  Please include you tooling setup eg. webpack config for webpack
  Please follow the guidelines for providing a minimal example: https://stackoverflow.com/help/mcve.
-->

Can be tested in the following repo: https://github.com/lasselupe33/wyw-in-js-next-repro

Can be tested by running `pnpm install` followed by
1. running `pnpm build` - Observe that the build fails here
2. running `pnpm build` again - Observe that the build now succeeds

--- a/.changeset/fix-next-turbopack-query-css.md
+++ b/.changeset/fix-next-turbopack-query-css.md
@@ -1,0 +1,6 @@
+---
+'@wyw-in-js/nextjs': patch
+'@wyw-in-js/turbopack-loader': patch
+---
+
+Fix Next.js Turbopack production builds by routing generated WyW CSS through a Turbopack query loader path on versions that support query conditions, while preserving sidecar output as the fallback for older Turbopack configs.

--- a/e2e/next-turbopack/package.json
+++ b/e2e/next-turbopack/package.json
@@ -4,7 +4,7 @@
   "dependencies": {
     "@wyw-in-js/nextjs": "workspace:*",
     "@wyw-in-js/template-tag-syntax": "workspace:*",
-    "next": "^16.0.0",
+    "next": "^16.2.0",
     "react": "^19.0.0",
     "react-dom": "^19.0.0"
   },

--- a/e2e/next-turbopack/scripts/test.mjs
+++ b/e2e/next-turbopack/scripts/test.mjs
@@ -1,4 +1,4 @@
-import { spawn } from 'node:child_process';
+import { execFileSync, spawn } from 'node:child_process';
 import { once } from 'node:events';
 import fs from 'node:fs/promises';
 import { createRequire } from 'node:module';
@@ -12,6 +12,7 @@ const require = createRequire(import.meta.url);
 const __filename = fileURLToPath(import.meta.url);
 const __dirname = path.dirname(__filename);
 const PKG_DIR = path.resolve(__dirname, '..');
+const MIN_NEXT_VERSION = { major: 16, minor: 2 };
 
 const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
 
@@ -125,6 +126,56 @@ const stopDevServer = async (child) => {
   ]);
 };
 
+const getNextBin = () => {
+  const nextPackageJson = require.resolve('next/package.json', {
+    paths: [PKG_DIR],
+  });
+
+  return path.resolve(path.dirname(nextPackageJson), 'dist', 'bin', 'next');
+};
+
+const getNextVersion = () => {
+  const nextPackageJson = require.resolve('next/package.json', {
+    paths: [PKG_DIR],
+  });
+  const { version } = require(nextPackageJson);
+  const [majorPart, minorPart] = version.split('.');
+
+  return {
+    major: Number.parseInt(majorPart, 10),
+    minor: Number.parseInt(minorPart, 10),
+    version,
+  };
+};
+
+const assertSupportedNextVersion = () => {
+  const { major, minor, version } = getNextVersion();
+  const isSupported =
+    major > MIN_NEXT_VERSION.major ||
+    (major === MIN_NEXT_VERSION.major && minor >= MIN_NEXT_VERSION.minor);
+
+  if (!isSupported) {
+    throw new Error(
+      `Expected Next.js ${MIN_NEXT_VERSION.major}.${MIN_NEXT_VERSION.minor}.x or newer, got ${version}`
+    );
+  }
+};
+
+const runProductionBuild = async () => {
+  await fs.rm(path.resolve(PKG_DIR, '.next'), { recursive: true, force: true });
+  await cleanupGeneratedCss();
+
+  execFileSync(process.execPath, [getNextBin(), 'build'], {
+    cwd: PKG_DIR,
+    env: {
+      ...process.env,
+      NODE_ENV: 'production',
+      NEXT_TELEMETRY_DISABLED: '1',
+    },
+    stdio: 'inherit',
+  });
+};
+
 const fetchWithRetries = async (url, child, logs) => {
   let lastError = null;
 
@@ -173,21 +224,40 @@ const runDevSmoke = async () => {
 };
 
 const readCssOutput = async () => {
-  const cssFile = path.resolve(
-    PKG_DIR,
-    'app',
-    'styles.wyw-in-js.module.css'
-  );
+  const cssFiles = [];
 
-  try {
-    return await fs.readFile(cssFile, 'utf8');
-  } catch (error) {
-    if (error?.code === 'ENOENT') {
-      throw new Error(`No generated CSS file found at ${cssFile}`);
+  const walk = async (dir) => {
+    let entries = [];
+    try {
+      entries = await fs.readdir(dir, { withFileTypes: true });
+    } catch (error) {
+      if (error?.code === 'ENOENT') return;
+      throw error;
     }
 
-    throw error;
+    await Promise.all(
+      entries.map(async (entry) => {
+        const fullPath = path.resolve(dir, entry.name);
+        if (entry.isDirectory()) {
+          await walk(fullPath);
+        } else if (entry.isFile() && entry.name.endsWith('.css')) {
+          cssFiles.push(fullPath);
+        }
+      })
+    );
+  };
+
+  await walk(path.resolve(PKG_DIR, '.next'));
+
+  if (cssFiles.length === 0) {
+    throw new Error('No CSS assets emitted by Next build');
   }
+
+  const contents = await Promise.all(
+    cssFiles.map((file) => fs.readFile(file, 'utf8'))
+  );
+
+  return contents.join('\n');
 };
 
 const assertMatches = (css, pattern, label) => {
@@ -198,9 +268,10 @@ const assertMatches = (css, pattern, label) => {
 
 const main = async () => {
   console.log(colors.blue('Package directory:'), PKG_DIR);
+  assertSupportedNextVersion();
 
   try {
-    await runDevSmoke();
+    await runProductionBuild();
 
     const cssOutput = await readCssOutput();
 
@@ -213,6 +284,8 @@ const main = async () => {
     assertMatches(cssOutput, /width:\s*200px/i, 'width');
     assertMatches(cssOutput, /font-size:\s*18px/i, 'font-size');
     assertMatches(cssOutput, /color:\s*tomato/i, 'color');
+
+    await runDevSmoke();
   } finally {
     await cleanupGeneratedCss();
   }

--- a/packages/nextjs/src/__tests__/withWyw.test.ts
+++ b/packages/nextjs/src/__tests__/withWyw.test.ts
@@ -1,22 +1,50 @@
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+
 import type { Configuration, RuleSetRule } from 'webpack';
 
 import { withWyw } from '../index';
 
 describe('withWyw', () => {
+  const withFakeNextVersion = <T>(version: string, callback: () => T): T => {
+    const previousCwd = process.cwd();
+    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wyw-next-'));
+    const nextPackageDir = path.join(tmpDir, 'node_modules', 'next');
+
+    fs.mkdirSync(nextPackageDir, { recursive: true });
+    fs.writeFileSync(
+      path.join(nextPackageDir, 'package.json'),
+      JSON.stringify({ version })
+    );
+
+    process.chdir(tmpDir);
+    try {
+      return callback();
+    } finally {
+      process.chdir(previousCwd);
+      fs.rmSync(tmpDir, { recursive: true, force: true });
+    }
+  };
+
   const getTurbopackLoaderOptions = (nextConfig: any) => {
     const rules =
       nextConfig.turbopack?.rules ?? nextConfig.experimental?.turbo?.rules;
     const tsRule = rules['*.ts'];
 
     if (Array.isArray(tsRule)) {
-      return tsRule[0].options;
+      const jsRule = tsRule.find(
+        (item) => item.loaders?.[0]?.options?.outputCss !== true
+      );
+
+      return jsRule?.loaders?.[0]?.options ?? tsRule[0].options;
     }
 
     return tsRule.loaders[0].options;
   };
 
-  it('injects Turbopack rules (turbopack.rules or experimental.turbo.rules)', () => {
-    const nextConfig = withWyw();
+  it('injects Turbopack query CSS rules for Next 16.2+', () => {
+    const nextConfig = withFakeNextVersion('16.2.4', () => withWyw());
 
     const rules =
       (nextConfig as any).turbopack?.rules ??
@@ -27,19 +55,49 @@ describe('withWyw', () => {
 
     const tsRule = rules['*.ts'];
 
-    if (Array.isArray(tsRule)) {
-      expect(tsRule[0].loader).toContain('turbopack-loader');
-      expect(tsRule[0].options.importOverrides).toMatchObject({
-        react: { mock: 'react' },
-      });
-      expect(tsRule[0].options).not.toHaveProperty('babelOptions');
-    } else {
-      expect(tsRule.loaders[0].loader).toContain('turbopack-loader');
-      expect(tsRule.loaders[0].options.importOverrides).toMatchObject({
-        react: { mock: 'react' },
-      });
-      expect(tsRule.loaders[0].options).not.toHaveProperty('babelOptions');
-    }
+    expect(Array.isArray(tsRule)).toBe(true);
+    expect(tsRule).toHaveLength(2);
+
+    const [cssRule, jsRule] = tsRule;
+    expect(jsRule.loaders[0].loader).toContain('turbopack-loader');
+    expect(jsRule.loaders[0].options.cssOutputMode).toBe('query');
+    expect(jsRule.loaders[0].options.importOverrides).toMatchObject({
+      react: { mock: 'react' },
+    });
+    expect(jsRule.loaders[0].options).not.toHaveProperty('babelOptions');
+    expect(jsRule.condition.all).toEqual(
+      expect.arrayContaining([
+        { not: 'foreign' },
+        { not: { query: expect.any(RegExp) } },
+      ])
+    );
+
+    expect(cssRule.loaders[0].loader).toContain('turbopack-loader');
+    expect(cssRule.loaders[0].options.cssOutputMode).toBe('query');
+    expect(cssRule.loaders[0].options.outputCss).toBe(true);
+    expect(cssRule.condition.all).toEqual(
+      expect.arrayContaining([
+        { not: 'foreign' },
+        { query: expect.any(RegExp) },
+      ])
+    );
+    expect(cssRule.as).toBe('*.module.css');
+  });
+
+  it('keeps the sidecar Turbopack rule for Next versions without query conditions', () => {
+    const nextConfig = withFakeNextVersion('16.1.1', () => withWyw());
+
+    const rules =
+      (nextConfig as any).turbopack?.rules ??
+      (nextConfig as any).experimental?.turbo?.rules;
+    const tsRule = rules['*.ts'];
+
+    expect(Array.isArray(tsRule)).toBe(false);
+    expect(tsRule.loaders[0].loader).toContain('turbopack-loader');
+    expect(tsRule.loaders[0].options.cssOutputMode).toBe('sidecar');
+    expect(tsRule.condition.all).toEqual(
+      expect.arrayContaining([{ not: 'foreign' }])
+    );
   });
 
   it('lets user-defined Turbopack rule keys win', () => {

--- a/packages/nextjs/src/index.ts
+++ b/packages/nextjs/src/index.ts
@@ -10,6 +10,8 @@ import type { NextConfig } from 'next';
 import type { Configuration, RuleSetRule, RuleSetUseItem } from 'webpack';
 
 const DEFAULT_EXTENSION = '.wyw-in-js.module.css';
+const CSS_OUTPUT_QUERY = '__wyw_css';
+const CSS_OUTPUT_QUERY_RE = new RegExp(`^\\??${CSS_OUTPUT_QUERY}$`);
 
 const DEFAULT_TURBO_RULE_KEYS = ['*.js', '*.jsx', '*.ts', '*.tsx'];
 
@@ -29,6 +31,10 @@ export type WywNextPluginOptions = {
 
 type NextWebpackConfigFn = NonNullable<NextConfig['webpack']>;
 type NextWebpackOptions = Parameters<NextWebpackConfigFn>[1];
+type NextVersion = {
+  major: number;
+  minor: number;
+};
 
 function isObject(value: unknown): value is Record<string, unknown> {
   return typeof value === 'object' && value !== null;
@@ -332,23 +338,48 @@ function injectWywLoader(
   ensureWywCssModuleRules(config, extension);
 }
 
-function shouldUseTurbopackConfig(nextConfig: NextConfig) {
-  const explicit = (nextConfig as unknown as Record<string, unknown>).turbopack;
-  if (typeof explicit !== 'undefined') {
-    return true;
-  }
-
+function getInstalledNextVersion(): NextVersion | null {
   try {
     const pkgPath = nodeRequire.resolve('next/package.json', {
       paths: [process.cwd()],
     });
     const pkg = nodeRequire(pkgPath) as { version?: unknown };
     const version = typeof pkg.version === 'string' ? pkg.version : '';
-    const major = Number.parseInt(version.split('.')[0] ?? '', 10);
-    return Number.isFinite(major) && major >= 16;
+    const [majorPart, minorPart] = version.split('.');
+    const major = Number.parseInt(majorPart ?? '', 10);
+    const minor = Number.parseInt(minorPart ?? '', 10);
+
+    if (!Number.isFinite(major) || !Number.isFinite(minor)) {
+      return null;
+    }
+
+    return { major, minor };
   } catch {
+    return null;
+  }
+}
+
+function shouldUseTurbopackConfig(
+  nextConfig: NextConfig,
+  nextVersion: NextVersion | null
+) {
+  const explicit = (nextConfig as unknown as Record<string, unknown>).turbopack;
+  if (typeof explicit !== 'undefined') {
+    return true;
+  }
+
+  return nextVersion !== null && nextVersion.major >= 16;
+}
+
+function supportsTurbopackQueryConditions(nextVersion: NextVersion | null) {
+  if (nextVersion === null) {
     return false;
   }
+
+  return (
+    nextVersion.major > 16 ||
+    (nextVersion.major === 16 && nextVersion.minor >= 2)
+  );
 }
 
 function injectWywTurbopackRules(
@@ -392,22 +423,58 @@ function injectWywTurbopackRules(
       : DEFAULT_REACT_IMPORT_OVERRIDES,
   };
 
-  const useTurbopackConfig = shouldUseTurbopackConfig(nextConfig);
+  const nextVersion = getInstalledNextVersion();
+  const useTurbopackConfig = shouldUseTurbopackConfig(nextConfig, nextVersion);
+  const useCssQueryOutput = supportsTurbopackQueryConditions(nextVersion);
+
+  const commonConditions = [
+    { not: 'foreign' },
+    { not: { path: /(?:^|[\\/])middleware\.[jt]sx?$/ } },
+  ];
+
+  const jsLoaderOptions = {
+    ...loaderOptions,
+    cssOutputMode: useCssQueryOutput ? 'query' : 'sidecar',
+  };
 
   const ruleValue = useTurbopackConfig
     ? {
-        loaders: [{ loader, options: loaderOptions }],
+        loaders: [{ loader, options: jsLoaderOptions }],
         condition: {
-          all: [
-            { not: 'foreign' },
-            { not: { path: /(?:^|[\\/])middleware\.[jt]sx?$/ } },
-          ],
+          all: commonConditions,
         },
       }
-    : [{ loader, options: loaderOptions }];
+    : [{ loader, options: jsLoaderOptions }];
+
+  const jsRuleValue = {
+    loaders: [{ loader, options: jsLoaderOptions }],
+    condition: {
+      all: [...commonConditions, { not: { query: CSS_OUTPUT_QUERY_RE } }],
+    },
+  };
+
+  const cssRuleValue = {
+    loaders: [
+      {
+        loader,
+        options: {
+          ...loaderOptions,
+          cssOutputMode: 'query',
+          outputCss: true,
+        },
+      },
+    ],
+    as: '*.module.css',
+    condition: {
+      all: [...commonConditions, { query: CSS_OUTPUT_QUERY_RE }],
+    },
+  };
 
   const wywRules = Object.fromEntries(
-    DEFAULT_TURBO_RULE_KEYS.map((key) => [key, ruleValue])
+    DEFAULT_TURBO_RULE_KEYS.map((key) => [
+      key,
+      useCssQueryOutput ? [cssRuleValue, jsRuleValue] : ruleValue,
+    ])
   );
 
   if (useTurbopackConfig) {

--- a/packages/turbopack-loader/src/__tests__/loader.test.ts
+++ b/packages/turbopack-loader/src/__tests__/loader.test.ts
@@ -21,7 +21,7 @@ describe('turbopack-loader', () => {
     transformMock.mockReset();
   });
 
-  it('writes CSS next to the module and injects an import after directives', async () => {
+  it('writes CSS next to the module and injects an import after directives by default', async () => {
     const { default: turbopackLoader } = await import('../index');
 
     const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wyw-turbo-'));
@@ -68,5 +68,95 @@ describe('turbopack-loader', () => {
 
     expect(emitted.code).toContain("'use client'");
     expect(emitted.code).toContain('import "./entry.wyw-in-js.module.css";');
+  });
+
+  it('injects a CSS query import in query output mode without writing a sidecar file', async () => {
+    const { default: turbopackLoader } = await import('../index');
+
+    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wyw-turbo-'));
+    const resourcePath = path.join(tmpDir, 'entry.tsx');
+    fs.writeFileSync(resourcePath, "'use client'\\nexport const x = 1;\\n");
+
+    transformMock.mockImplementation(async (_services, code) => {
+      return {
+        code,
+        sourceMap: null,
+        cssText: '.a{color:red}',
+        dependencies: [],
+      };
+    });
+
+    const emitted: { code?: string } = {};
+
+    await new Promise<void>((resolve, reject) => {
+      turbopackLoader.call(
+        {
+          addDependency: jest.fn(),
+          async: jest.fn(),
+          callback: (err: Error | null, code?: string) => {
+            if (err) reject(err);
+            else {
+              emitted.code = code;
+              resolve();
+            }
+          },
+          emitWarning: jest.fn(),
+          getOptions: () => ({ cssOutputMode: 'query' }),
+          getResolve: () => async () => false,
+          resourcePath,
+        } as any,
+        fs.readFileSync(resourcePath, 'utf8'),
+        null
+      );
+    });
+
+    const cssFilePath = path.join(tmpDir, 'entry.wyw-in-js.module.css');
+    expect(fs.existsSync(cssFilePath)).toBe(false);
+
+    expect(emitted.code).toContain("'use client'");
+    expect(emitted.code).toContain('import "./entry.tsx?__wyw_css";');
+  });
+
+  it('returns CSS for the CSS query loader branch', async () => {
+    const { default: turbopackLoader } = await import('../index');
+
+    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'wyw-turbo-'));
+    const resourcePath = path.join(tmpDir, 'entry.tsx');
+    fs.writeFileSync(resourcePath, 'export const x = 1;\n');
+
+    transformMock.mockImplementation(async (_services, code) => {
+      return {
+        code,
+        sourceMap: null,
+        cssText: '.a{color:red}',
+        dependencies: [],
+      };
+    });
+
+    const emitted: { code?: string } = {};
+
+    await new Promise<void>((resolve, reject) => {
+      turbopackLoader.call(
+        {
+          addDependency: jest.fn(),
+          async: jest.fn(),
+          callback: (err: Error | null, code?: string) => {
+            if (err) reject(err);
+            else {
+              emitted.code = code;
+              resolve();
+            }
+          },
+          emitWarning: jest.fn(),
+          getOptions: () => ({ outputCss: true }),
+          getResolve: () => async () => false,
+          resourcePath,
+        } as any,
+        fs.readFileSync(resourcePath, 'utf8'),
+        null
+      );
+    });
+
+    expect(emitted.code).toBe(':global(.a){color:red}');
   });
 });

--- a/packages/turbopack-loader/src/index.ts
+++ b/packages/turbopack-loader/src/index.ts
@@ -13,6 +13,7 @@ import { writeFileIfChanged } from './file-utils';
 import { insertImportStatement } from './insert-import';
 
 const DEFAULT_EXTENSION = '.wyw-in-js.module.css';
+const CSS_OUTPUT_QUERY = '__wyw_css';
 
 const stripQueryAndHash = (request: string) => {
   const queryIdx = request.indexOf('?');
@@ -27,7 +28,9 @@ const stripQueryAndHash = (request: string) => {
 };
 
 export type LoaderOptions = {
+  cssOutputMode?: 'sidecar' | 'query';
   keepComments?: boolean;
+  outputCss?: boolean;
   prefixer?: boolean;
   sourceMap?: boolean;
 } & Partial<PluginOptions>;
@@ -100,8 +103,15 @@ const turbopackLoader: Loader = function turbopackLoader(
 
   logger('turbopack-loader %s', this.resourcePath);
 
-  const { sourceMap, keepComments, prefixer, configFile, ...rest } =
-    this.getOptions() || {};
+  const {
+    sourceMap,
+    keepComments,
+    outputCss,
+    cssOutputMode = 'sidecar',
+    prefixer,
+    configFile,
+    ...rest
+  } = this.getOptions() || {};
 
   if (configFile) {
     const configPath = path.isAbsolute(configFile)
@@ -176,16 +186,33 @@ const turbopackLoader: Loader = function turbopackLoader(
           )
         );
 
-        writeFileIfChanged(cssFilePath, cssText);
+        if (outputCss) {
+          callback(null, cssText);
+          return;
+        }
 
-        const importStatement = `import ${JSON.stringify(cssImportPath)};`;
+        let importPath = cssImportPath;
+        if (cssOutputMode === 'query') {
+          importPath = `./${path.basename(
+            this.resourcePath
+          )}?${CSS_OUTPUT_QUERY}`;
+        } else {
+          writeFileIfChanged(cssFilePath, cssText);
+        }
+
+        const importStatement = `import ${JSON.stringify(importPath)};`;
         const finalCode = insertImportStatement(result.code, importStatement);
 
         callback(null, finalCode, result.sourceMap ?? undefined);
         return;
       }
 
-      if (fs.existsSync(cssFilePath)) {
+      if (outputCss) {
+        callback(null, '');
+        return;
+      }
+
+      if (cssOutputMode !== 'query' && fs.existsSync(cssFilePath)) {
         writeFileIfChanged(cssFilePath, '');
       }
 

